### ISSUE TITLE
Discriminator prediction values exceed 1.0 when compared to binary true/false label

### ISSUE THREAD

**Issue by urimerhav** [NVIDIA/pix2pixHD#126]
This issue relates specifically to discriminator outputs are used to calculate `loss_D_real`.

It ultimately boils down to the code minimizing `||pred_real-1.0||`. 

This would make sense of pred_real was upper-bound to be 1.0, for example by taking `sigmoid(pred_real)`. The way the code is currently laid out, the discriminator has a weird job where it has to predict bigger numbers when the photo is real, but never bigger than 1.0. 

I don't think I found a mention of this in the article: is it intentional or a (potentially not very harmful) bug?

**Comment by hannesliik** [comment]
It is the LSGAN objective function and intentional. I think the intuition is that it penalizes the discriminator for being too confident, which is supposed to make it easier for the generator to learn. If you specify the "vanilla" objective, IIRC, the sigmoid is hidden in the loss function.

**Comment by aayushARM** [comment]
@urimerhav @hannesliik Hello both. Thanks for clarifying, I had similar confusion. However, I am still not able to make complete sense of the tensorboard loss curves produced by this model. Could you people please throw some insight and point me in the right direction?
So for my run on a custom dataset, I could get pretty decent looking 300x300 samples. The tensorboard training graphs are as below:

Discriminator loss for Real Samples:
![D_real](https://user-images.githubusercontent.com/20727454/62172188-6f95be80-b2ff-11e9-9a06-8a57feeea19e.png)

Discriminator loss for Generated Samples:
![D_fake](https://user-images.githubusercontent.com/20727454/62172196-7b818080-b2ff-11e9-9bfb-533f2488e8e5.png)

Generator Least-Squares GAN Loss:
![G_GAN](https://user-images.githubusercontent.com/20727454/62172270-b4215a00-b2ff-11e9-96d3-03d37b142f78.png)

Generator Feature Matching Loss(from Discriminator layers):
![G_GAN_Feat](https://user-images.githubusercontent.com/20727454/62172568-a28c8200-b300-11e9-86fc-0a1d3d55dc6b.png)

Optional Generator VGG Perceptual Loss:
![G_VGG](https://user-images.githubusercontent.com/20727454/62172578-b33cf800-b300-11e9-9592-4e223879776f.png)

I am still a beginner at GANs, and from what theory I know about a typical cross-entropy conditional GAN loss is that during training, the loss of D({label, real}) and D({label, fake}) should show a trend of increase upto some point(as G gets better and desired values for both move away from 1 and 0 respectively towards 0.5), and the G loss should show a trend of decrease upto some point(as G gets better and desired value gets closer to 1 starting from 0). But what's happening here is a bit different. Any ideas how should I be interpreting these curves?

**Comment by hannesliik** [comment]
IMO GAN losses are not interpretable, only the generated images are. Of course, I am not a GAN expert either.
